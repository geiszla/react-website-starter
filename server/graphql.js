import {
  GraphQLBoolean,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString
} from 'graphql';

import bcrypt from 'bcrypt-nodejs';

// Queries
const queryType = new GraphQLObjectType({
  name: 'Query',
  fields: {
    getUsername: {
      type: GraphQLString,
      resolve: ({ session }) => {
        if (session.isLoggedIn) return session.username;
        return null;
      }
    }
  }
});

// Mutations
const mutationType = new GraphQLObjectType({
  name: 'Mutation',
  fields: {
    loginUser: {
      type: new GraphQLNonNull(GraphQLBoolean),
      args: {
        username: { type: new GraphQLNonNull(GraphQLString) },
        password: { type: new GraphQLNonNull(GraphQLString) }
      },
      resolve: ({ session }, { username, password }) => new Promise((resolve) => {
        const correctPassword = '$2a$08$7OBLs4B/PZvSxafCdjIE8.qost5k7QSyS7tUlgPi0ckDpOYYNAhC.'; // testpass

        bcrypt.compare(password, correctPassword, (err, res) => {
          if (err) {
            console.error('Couldn\'t compare password hases.');
            resolve(false);
          }

          if (res === true) {
            session.isLoggedIn = true;
            session.username = username;

            resolve(session.isLoggedIn);
          } else {
            resolve(false);
          }
        });
      })
    },
    logoutUser: {
      type: new GraphQLNonNull(GraphQLBoolean),
      resolve: ({ session }) => {
        session.isLoggedIn = false;
        return session.isLoggedIn;
      }
    }
  }
});

export default new GraphQLSchema({ query: queryType, mutation: mutationType });
