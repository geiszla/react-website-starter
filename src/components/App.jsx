/* eslint react/no-unused-state: 0 */

import gql from 'graphql-tag';
import { observable } from 'mobx';
import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import React, { Component, Fragment } from 'react';
import { compose, graphql } from 'react-apollo';
import { hot } from 'react-hot-loader';
import {
  Redirect,
  Route,
  Switch,
  withRouter
} from 'react-router-dom';

import { withStyles } from '@material-ui/core/styles';

import { Home, Login } from './routes.jsx';

const styles = () => ({
  background: {
    backgroundPosition: '35%',
    backgroundRepeat: 'no-repeat',
    backgroundSize: 'cover',
    filter: 'blur(10px)',
    height: '100%',
    position: 'fixed',
    transition: 'opacity 500ms ease-in-out',
    width: '100%'
  }
});

@observer
class App extends Component {
  @observable usernameError = false;

  @observable passwordError = false;

  fromLogin = false;

  componentDidMount = () => {
    const jssStyles = document.getElementById('jss-server-side');
    if (jssStyles && jssStyles.parentNode) {
      jssStyles.parentNode.removeChild(jssStyles);
    }
  }

  handleLogin = async () => {
    const username = document.getElementById('username').value;
    const password = document.getElementById('password').value;

    if (!username) {
      this.passwordError = false;
      this.usernameError = true;
      return;
    }

    this.usernameError = false;

    const { data } = await this.props.loginMutate({
      variables: { username, password }
    });

    if (data.loginUser === true) {
      this.passwordError = false;
      this.fromLogin = true;
    } else {
      this.passwordError = true;
    }
  }

  handleLogout = () => {
    this.props.logoutMutate();
  }

  render() {
    const { passwordError, usernameError } = this;
    const { pathname } = this.props.location;
    const username = this.props.data.getUsername;

    const isLoggedIn = username !== null;

    if (!isLoggedIn && pathname !== '/login') {
      return <Redirect push to="/login" />;
    }

    if (isLoggedIn === true && pathname === '/login') {
      return <Redirect push to="/" />;
    }

    return (
      <Fragment>
        <div
          className={this.props.classes.background}
          style={{ backgroundImage: pathname === '/login' || this.fromLogin ? 'url(images/out.jpg)' : 'url(images/in.jpg)' }}
        />
        <Switch>
          <Route
            path="/login"
            render={defaultProps => (
              <Login
                handleLogin={this.handleLogin}
                usernameError={usernameError}
                passwordError={passwordError}
                {...defaultProps}
              />
            )}
          />
          <Route
            path="/"
            render={defaultProps => (
              <Home
                handleLogout={this.handleLogout}
                username={username}
                fromLogin={this.fromLogin}
                {...defaultProps}
              />
            )}
          />
        </Switch>
      </Fragment>
    );
  }
}

const getUsernameQuery = gql`
  query getUsernameQuery {
    getUsername
  }
`;

const LoginUserMutation = gql`
  mutation LoginUserMutation($username: String!, $password: String!) {
    loginUser(username: $username, password: $password)
  }
`;

const LogoutUserMutation = gql`
  mutation LogoutUserMutation {
    logoutUser
  }
`;

const loginMutationOptions = {
  refetchQueries: ['getUsernameQuery', 'FetchStatusQuery']
};

App.propTypes = {
  classes: PropTypes.shape({
    background: PropTypes.string.isRequired
  }).isRequired,
  data: PropTypes.shape({
    getUsername: PropTypes.string
  }).isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired
  }).isRequired,
  loginMutate: PropTypes.func.isRequired,
  logoutMutate: PropTypes.func.isRequired
};

export default compose(
  hot(module),
  graphql(getUsernameQuery),
  graphql(LoginUserMutation, {
    name: 'loginMutate',
    options: loginMutationOptions
  }),
  graphql(LogoutUserMutation, {
    name: 'logoutMutate',
    options: loginMutationOptions
  }),
  withRouter,
  withStyles(styles)
)(App);
