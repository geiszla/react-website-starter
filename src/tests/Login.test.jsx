import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloClient } from 'apollo-client';
import { SchemaLink } from 'apollo-link-schema';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { printSchema } from 'graphql';
import { makeExecutableSchema } from 'graphql-tools';
import React from 'react';
import { ApolloProvider } from 'react-apollo';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';

import schema from '../../server/graphql';
import App from '../components/App.jsx';

// Helper functions
function testRoute(path) {
  const tree = renderer.create(
    <ApolloProvider client={apolloClient}>
      <MemoryRouter initialEntries={[path]}>
        <App />
      </MemoryRouter>
    </ApolloProvider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
}

// Setup
const executableSchema = makeExecutableSchema({
  typeDefs: printSchema(schema),
  resolverValidationOptions: {
    requireResolversForResolveType: false
  }
});

const apolloClient = new ApolloClient({
  link: new SchemaLink({ schema: executableSchema }),
  cache: new InMemoryCache()
});

configure({ adapter: new Adapter() });
it('renders the loading component when different routes are requested', () => {
  console.error = undefined;
  testRoute('/login');
  testRoute('/');
});
