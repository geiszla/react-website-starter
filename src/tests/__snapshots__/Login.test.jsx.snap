// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the loading component when different routes are requested 1`] = `
Array [
  <div
    className="App-background-1"
    style={
      Object {
        "backgroundImage": "url(images/in.jpg)",
      }
    }
  />,
  <div>
    Loading...
  </div>,
]
`;

exports[`renders the loading component when different routes are requested 2`] = `
Array [
  <div
    className="App-background-1"
    style={
      Object {
        "backgroundImage": "url(images/in.jpg)",
      }
    }
  />,
  <div>
    Loading...
  </div>,
]
`;
